### PR TITLE
[Windows, Keyboard] Fix crash on Up-only IME event

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR fixes https://github.com/flutter/flutter/issues/104169, where a specific combination of environments leads to the following key events:
- KeyDown, ScanCode ShiftLeft, VirtualKey ShiftLeft.
- KeyUp, ScanCode ShiftLeft, VirtualKey VK_PROCESSKEY.

Besides fixing the exact part, this PR also reorganized much of the synthesization algorithm, adopting the terminology and process from the latest version used in [Android](https://github.com/flutter/engine/pull/33686). 

### Cause

The previous algorithm always looks for the critical key by virtual key, and thus fails to consider this event as the target key, synthesizing a key up event. However, the "main event" stage uses the pressing records _before_ the synthesization, therefore also dispatching a key up event.

### What is changed

The most important parts of the change are as follows:

1. The synthesization algorithm now looks for the critical key by physical key if the key is pressed. In this way, the extra key up event are no longer synthesized. https://github.com/flutter/engine/pull/33746/files#diff-d948a303b1d747955e0f2b1e73bd03b8b4d103fd1dfda7df97dc620603e08c8aR431-R442

2. The information for the main event is computed _after_ synthesization. In this way, even if the synthesization is incorrect, the main-event stage won't exacerbate the problem. https://github.com/flutter/engine/pull/33746/files#diff-d948a303b1d747955e0f2b1e73bd03b8b4d103fd1dfda7df97dc620603e08c8aR171-R208

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/keyboard_key_embedder_handler.h [hunk 1]
```diff
@@ -114,10 +114,13 @@ class KeyboardKeyEmbedderHandler
                                 uint64_t logical_key);
   // Check each key's state from |get_key_state_| and synthesize events
   // if their toggling states have been desynchronized.
-  void SynchronizeCritialToggledStates(int virtual_key, bool is_down);
+  void SynchronizeCritialToggledStates(int this_virtual_key,
```
Comment (gspencergoog) [inline shell/platform/windows/keyboard_key_embedder_handler.h:117]: I don't think adding "this_" helps explain anything.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: shell/platform/windows/keyboard_key_embedder_handler.h (hunk 1 above)
Comment (dkwingsmt) [reply shell/platform/windows/keyboard_key_embedder_handler.h]: You're right. It is the leftover of an intermediate state. I've changed them to be `event_virtual_key` and alike, which should be more helpful.

**Comment by chinmaygarde** [comment]
Not sure how the presubmit is running for 12 hours. I can't get to the status page either. cc @godofredoc

**Comment by godofredoc** [comment]
> Not sure how the presubmit is running for 12 hours. I can't get to the status page either. cc @godofredoc

@drewroengoogle @yusuf-goog can you please take a look to identify if the task never run or if it was github not updating the check? some instructions on how to triage the problem can be found here: go/flutter-luci-playbook#debug-builds-not-being-scheduled

**Comment by drewroengoogle** [comment]
> > Not sure how the presubmit is running for 12 hours. I can't get to the status page either. cc @godofredoc
> 
> @drewroengoogle @yusuf-goog can you please take a look to identify if the task never run or if it was github not updating the check? some instructions on how to triage the problem can be found here: go/flutter-luci-playbook#debug-builds-not-being-scheduled

If I'm understanding this correctly, it does look like this run did fail: https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20iOS%20Engine/32658/overview

**Comment by yusuf-goog** [comment]
Kicking off a retry to see if that reports back.

**Comment by godofredoc** [comment]
> If I'm understanding this correctly, it does look like this run did fail: https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20iOS%20Engine/32658/overview

@dkwingsmt to validate the first build failed and after clicking the re-run button the check was not updated, am I correct?

@yusuf-goog @drewroengoogle any 500s in the reset-try api?

**Pull request merged**